keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,430 @@
+
+#### Container Restart Policy
+
+To provide some fine-grained control to the user,
+PodSpec.Container.ResourceRequirements is extended with ResizePolicy flag
+for each resource type (CPU, memory) :
+* NoRestart - the default value; resize the Container without restarting it,
+* RestartContainer - restart the Container in-place to apply new resource
+  values (e.g. Java process needs to change its Xmx flag),
+* RestartPod - restart the whole Pod in-place to apply new resource values
+  (e.g. Pod requires its Init Containers to re-run).
+
+By using the ResizePolicy flag, user can mark Containers or Pods as safe
+(or unsafe) for in-place resource update.
+
+This flag is used by Kubelet to determine the actions needed. This flag **may** be
+used by the actors starting the update to decide if the process should be started
+at all (for example VPA might decide to evict Pod with RestartPod policy).
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problems whereas
+changes to available memory are more probable to require restarts.
+
+If more than one resource type with different policies are updated, then
+RestartPod policy takes precedence over RestartContainer, which in turn takes
+precedence over NoRestart policy.
+
+#### Pod Resize Retry Policy
+
+If resource update fails, say due to lack of space on Node, default behavior
+is to let the initiating actor such as VPA handle the failure. Alternately, a
+Controller can either retry the update, or reschedule the Pod based on policy.
+
+PodSpec is extended with a new flag, PodSpec.RetryPolicy, with possible values:
+* NoRetry - the default value; do nothing, initiating actor handles failure,
+* RetryUpdate - Controller retries resource update in-place,
+* Reschedule - Controller evicts Pod, and creates updated Pod for scheduling.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter is changed to *runtimeapi.ContainerResources*,
+that is runtime agnostic.
+
+### Flow Control
+
+Presently, when a Pod is created, Scheduler is responsible for selecting a suitable
+Node that accommodates the Pod. In order to allow Pod resizing operation to work
+well with new Pod scheduling workflow, this design keeps Scheduler in the Pod
+resizing decision path.
+
+The following steps denote a typical flow of an in-place resize operation for a Pod
+with ResizePolicy set to Update for all its Containers.
+
+1. The initiating actor updates ResourceRequirements using PATCH verb.
+1. API Server validates the new ResourceRequirements
+   (e.g. limits are not below requested resources, QoS class does not change).
+1. API Server calls all Admission Controllers to verify the Pod Update.
+   * If any of the Controllers reject the update, they set PodResizeSuccess
```

Comment:
if an admission controller rejects the update, the update should not be be made, and there is no need to write anything to the pod spec.  the controller that is attempting to make the update needs to respond to the update failure and backoff.

Reply:
The condition is not there anymore, resolving this issue.